app/util/rpc-domain-utils.test.ts

```diff
@@ -1,152 +1,377 @@
+        // Exercise
+        const result = await getSafeChainsListFromCacheOnly();
+        // Verify
+        expect(result).toEqual([]);
+      });
     });
-
-    it('should return false for unknown domains', () => {
-      expect(isKnownDomain('unknown.com')).toBe(false);
-      expect(isKnownDomain('custom.net')).toBe(false);
+    it('handles JSON parse errors gracefully', async () => {
+      // Mock StorageWrapper to return invalid JSON
+      (StorageWrapper.getItem as jest.Mock).mockResolvedValueOnce('invalid-json');
+      const result = await getSafeChainsListFromCacheOnly();
+      expect(result).toEqual([]);
     });
-
-    it('should be case-insensitive', () => {
-      expect(isKnownDomain('EXAMPLE.COM')).toBe(true);
-      expect(isKnownDomain('Infura.Io')).toBe(true);
+    it('handles storage errors gracefully', async () => {
+      // Mock StorageWrapper to throw an error
+      (StorageWrapper.getItem as jest.Mock).mockRejectedValueOnce(new Error('Storage error'));
+      const result = await getSafeChainsListFromCacheOnly();
+      expect(result).toEqual([]);
     });
   });
-
-  describe('getNetworkRpcUrl', () => {
-    beforeEach(() => {
-      jest.clearAllMocks();
+  describe('initializeRpcProviderDomains', () => {
+    describe('when chains list contains valid RPC URLs', () => {
+      it('initializes known domains from the chains list', async () => {
+        // Setup
+        resetModuleState(); // Explicitly reset state
+        const mockChains: SafeChain[] = [
+          {
+            chainId: '1',
+            name: 'Ethereum',
+            nativeCurrency: { symbol: 'ETH' },
+            rpc: ['https://mainnet.infura.io', 'https://eth-mainnet.alchemyapi.io'],
+          },
+        ];
+        (StorageWrapper.getItem as jest.Mock).mockResolvedValue(JSON.stringify(mockChains));
+        // Exercise
+        await initializeRpcProviderDomains();
+        // Verify
+        const knownDomains = getKnownDomains();
+        expect(knownDomains).toBeInstanceOf(Set);
+        expect(knownDomains?.has('mainnet.infura.io')).toBe(true);
+        expect(knownDomains?.has('eth-mainnet.alchemyapi.io')).toBe(true);
+      });
     });
-
-    it('should return the RPC URL when network client and configuration are found', () => {
-      // Mock implementation
-      const mockNetworkClientId = 'mock-network-client-id';
-      (Engine.context.NetworkController.findNetworkClientIdByChainId as jest.Mock).mockReturnValue(mockNetworkClientId);
-      (Engine.context.NetworkController.getNetworkConfigurationByNetworkClientId as jest.Mock).mockReturnValue({
-        rpcEndpoints: [
-          { url: 'https://mainnet.infura.io/v3/123' },
-          { url: 'https://backup.example.com' },
-        ],
-        defaultRpcEndpointIndex: 0,
+    describe('when chains list contains invalid RPC URLs', () => {
+      it('handles invalid URLs gracefully', async () => {
```

Comment:
S: be more specific and avoid "handles" and "gracefuly".
What are your definitions of that here?
I see you are expecting knownDomains:
- to be a set
- to contain item "mainnet.infura.io"
- to only have one item.

From that I deduce that you expect that the invalid item was not added.

```suggestion
      it('does not add invalid URLs from knownDomains', async () => {
```